### ISSUE TITLE
Misaligned labels

### ISSUE THREAD

**Issue by leobasilio** [ajrogers17/Forex_Prediction#1]
Hi. First of all, thank you for publishing your code. I've been studying it for a while, and unfortunately there's a mistake in it. When slicing the data into windows, the label associated with each window is the one from its first element, not the one from its last element. That's not what we want to predict, and it also allows the neural network to easily converge, as the solution lies inside the window itself. It's just the difference between the prices of the first and the second element. That's why we get such a high accuracy metric. After fixing that, the accuracy stays around 50%.

**Comment by paul22** [comment]
> Hi. First of all, thank you for publishing your code. I've been studying it for a while, and unfortunately there's a mistake in it. When slicing the data into windows, the label associated with each window is the one from its first element, not the one from its last element. That's not what we want to predict, and it also allows the neural network to easily converge, as the solution lies inside the window itself. It's just the difference between the prices of the first and the second element. That's why we get such a high accuracy metric. After fixing that, the accuracy stays around 50%.

Hi
I'm new to Machine Learning and Python.
In your comment are you referring to split the test and training data, e.g. using train_test_split from Sklearn?
Thanks

**Comment by leobasilio** [comment]
Hi. I'm referring to the author's own method. Before splitting the data, we have this:

| i | mcad | sma | ... | price | label |
|:-:|:----:|-----|-----|-------|-------|
| 0 |  ... | ... | ... | ...   | 0     |
| 1 |  ... | ... | ... | ...   | 1     |
| 2 |  ... | ... | ...  | ...   | 1     |

For each example, `label = 0` if the next price is lower than the current price and `label = 1` otherwise. So far so good. We want to predict the label and the label tells us something about the future.

Now take a look at this part:

```python3
X_train, _ = split_sequences(train, bin[:test_range], 32)
Y_train = bin[:test_range]
```

`split_sequences` will break our data frame into sliding windows, and `bin` contains the labels for each example, but now suddenly each label is associated with a window, not an example. So we have this:

|`bin[0]`| i = | 0 | 1 | 2 | 3 | 4 | ... | k | 
|---|---|---|---|---|---|---|---|---|

|`bin[1]` | i = | 1 | 2 | 3 | 4 | 5 | ... | k+1|
|---|---|---|---|---|---|---|---|---|

|`bin[2]`| i = | 2 | 3 | 4 | 5 | 6 | ... | k+2|
|---|---|---|---|---|---|---|---|---|

Now the label for the first window is the same label for the first example. The label for the second window is the same label for the second example, and so on. Because we're using sliding windows, that label will always be the label for the first example in the window. Remember that the label is based on the next example? The next example is inside the window, and it's always the second element. So for every window the calculation for predicting the label is pretty straightforward. We just need to compare the second and the first example inside it.

That's not what we want. We want that, given a window, we are able to predict the next price movement out of this window, which is the same as predicting the label for the last element in the window, not the first one.
